File: peer-review/TTK4145-2025-code-peer-review-files/88a7def0/models/elevator.go
package models

type ElevatorState struct {
	Id        Id
	Floor     int
	Behavior  ElevatorBehavior
	Direction MotorDirection
}

type ElevatorBehavior int

const (
	Idle ElevatorBehavior = iota
	DoorOpen
	Moving
)

type Orders [][3]bool

type MotorDirection int

const (
	Up   MotorDirection = 1
	Down                = -1
	Stop                = 0
)

type ButtonType int

const (
	HallUp   ButtonType = 0
	HallDown            = 1
	Cab                 = 2
)

File: peer-review/TTK4145-2025-code-peer-review-files/b67214ff/single_elevator/src/execution/elevator.rs
use std::time::Duration;

pub const N_FLOORS :usize = 4;
pub const N_BUTTONS :usize = 3;


#[derive(Copy, Clone, Debug, PartialEq,serde::Deserialize,serde::Serialize)]
#[repr(u8)]
pub enum Dirn { 
    #[serde(rename = "down")]
    Down = u8::MAX,
    #[serde(rename = "stop")]
    Stop = 0,
    #[serde(rename = "up")]
    Up = 1
}

#[derive(Copy, Clone, Debug, PartialEq, num_derive::FromPrimitive)]
#[repr(u8)]
pub enum Button { 
    HallUp,
    HallDown,
    Cab
}


#[repr(u8)]
#[derive(Copy, Clone, Debug, PartialEq,serde::Deserialize,serde::Serialize)]
pub enum ElevatorBehaviour {
    #[serde(rename = "idle")]
    Idle,
    #[serde(rename = "doorOpen")]
    DoorOpen,
    #[serde(rename = "moving")]
    Moving
}

#[derive(Clone, Copy)]
pub struct Elevator {
    pub floor : Option<u8>, 
    pub dirn : Dirn,
    pub requests : [[bool; N_BUTTONS];  N_FLOORS],
    pub behaviour : ElevatorBehaviour,
    pub door_open_duration_s : Duration
}


impl Elevator {
pub fn elevator_init() -> Elevator {
    Elevator{
        floor : None,
        dirn : Dirn::Stop,
        behaviour : ElevatorBehaviour::Idle,
        requests: [[false; N_BUTTONS];N_FLOORS],
        door_open_duration_s : Duration::from_secs(3) // Door open timer is set here
    }
}
}
